src/utils/match/test.ts

```diff
@@ -0,0 +1,91 @@
+import { match } from "."
+import { Maybe, MaybeTag, Some, None } from "../../types/Maybe"
+
+interface User {
+  name: string
+}
+
+describe("match", () => {
+  describe("with all cases", () => {
+    // Arrange
+    const matchUser = match<Maybe<User>, string>({
+      [MaybeTag.None]: x => x.type,
+      [MaybeTag.Some]: x => x.data.name,
+    })
+
+    it("matches a Tagged member", () => {
+      // Arrange
+      const user = Some({ name: "Rob" })
+
+      // Assert
+      expect(matchUser(user)).toEqual(user.data.name)
+    })
+  })
+
+  describe("with caseDefault", () => {
+    // Arrange
+    const matchUserWithDefault = match<Maybe<User>, string>(
+      {
+        [MaybeTag.Some]: x => x.data.name,
+      },
+      () => "This user doesn't exist",
+    )
+
+    it("handles non-matching with caseDefault", () => {
+      // Arrange
+      const user = None
+
+      // Assert
+      expect(matchUserWithDefault(user)).toEqual("This user doesn't exist")
+    })
+
+    it("handles undefined with caseDefault", () => {
+      // Arrange
+      const user = undefined
+
+      // Assert
+      expect(matchUserWithDefault(user)).toEqual("This user doesn't exist")
+    })
+  })
+
+  describe("massive misuse", () => {
```

Comment:
Hard to think outside the types on this one, but technically the function can return undefined—but only if you abandon the type system...which is kind of the whole point.